test/jdk/java/awt/regtesthelpers/PassFailJFrame.java

```diff
@@ -54,101 +65,156 @@
+                timeout = true;
+                latch.countDown();
+            }
+            testTimeoutLabel.setText(String.format("Test timeout: %s", convertMillisToTimeStr(leftTime)));
+        });
+        timer.start();
+        frame.add(testTimeoutLabel, BorderLayout.NORTH);
+        frame.add(new JScrollPane(instructionsText), BorderLayout.CENTER);
 
         JButton btnPass = new JButton("Pass");
-        btnPass.addActionListener((e) -> latch.countDown());
+        btnPass.addActionListener((e) -> {
+            latch.countDown();
+            timer.stop();
+        });
 
         JButton btnFail = new JButton("Fail");
-        btnFail.addActionListener((e) -> getFailureReason());
+        btnFail.addActionListener((e) -> {
+            getFailureReason();
+            timer.stop();
+        });
 
         JPanel buttonsPanel = new JPanel();
         buttonsPanel.add(btnPass);
         buttonsPanel.add(btnFail);
 
+        frame.addWindowListener(new WindowAdapter() {
+            @Override
+            public void windowClosing(WindowEvent e) {
+                super.windowClosing(e);
+                testFailedReason = "Failure Reason:\n User closed the " +
+                        "instruction Frame";
+                failed = true;
+                latch.countDown();
+            }
+        });
+
         frame.add(buttonsPanel, BorderLayout.SOUTH);
         frame.pack();
-        frame.setLocation(10, 10);
+        frame.setLocationRelativeTo(null);
         frame.setVisible(true);
+
+        frameList.add(frame);
+    }
+
+    private static String convertMillisToTimeStr(long millis) {
+        if (millis < 0) {
+            return "00:00:00";
+        }
+        long hours = millis / 3600000;
+        long minutes = (millis - hours * 3600000) / 60000;
+        long seconds = (millis - hours * 3600000 - minutes * 60000) / 1000;
+        return String.format("%02d:%02d:%02d", hours, minutes, seconds);
     }
 
     /**
      * Wait for the user decision i,e user selects pass or fail button.
      * If user does not select pass or fail button then the test waits for
      * the specified timeoutMinutes period and the test gets timeout.
+     * Note: This method should be called from main() thread
      *
-     * @throws InterruptedException
-     * @throws InvocationTargetException
+     * @throws InterruptedException      exception thrown for invokeAndWait
+     * @throws InvocationTargetException exception thrown for invokeAndWait
      */
     public void awaitAndCheck() throws InterruptedException, InvocationTargetException {
-        boolean timeoutHappened = !latch.await(this.timeoutMinutes,
-                TimeUnit.MINUTES);
+        latch.await();
         if (isEventDispatchThread()) {
-            disposePassFailJFrame();
-        } else invokeAndWait(() -> disposePassFailJFrame());
+            disposeFrames();
+        } else invokeAndWait(PassFailJFrame::disposeFrames);
```

Comment:
This is bad from Java Code Style point of view.

Additionally, you stated this method is not to be called on EDT, therefore you know `disposeFrames` should be called via `invokeAndWait`.

Shall this method throw an exception if it's called on EDT? It will prevent accidental mistakes.